src/network/ClientConnectionManager.ts

```diff
@@ -0,0 +1,804 @@
+
+    public getOrConnect(address: Address): Promise<ClientConnection> {
+        if (!this.client.getLifecycleService().isRunning()) {
+            return Promise.reject(new ClientNotActiveError('Client is not active.'));
+        }
+
+        const connection = this.getConnectionFromAddress(address);
+        if (connection) {
+            return Promise.resolve(connection);
+        }
+
+        const addressKey = address.toString();
+        const pendingConnection = this.pendingConnections.get(addressKey);
+        if (pendingConnection) {
+            return pendingConnection.promise;
+        }
+
+        const connectionResolver: Promise.Resolver<ClientConnection> = DeferredPromise<ClientConnection>();
+        this.pendingConnections.set(addressKey, connectionResolver);
+
+        const processResponseCallback = (msg: ClientMessage) => {
+            this.client.getInvocationService().processResponse(msg);
+        };
+
+        this.translate(address)
+            .then((translatedAddress) => {
+                if (translatedAddress == null) {
+                    connectionResolver.reject(new RangeError(`Address Translator could not translate address ${address}`));
+                    return;
+                }
+
+                this.triggerConnect(translatedAddress)
+                    .then((socket) => {
+                        const clientConnection = new ClientConnection(this.client, translatedAddress, socket,
+                            this.connectionIdCounter++);
+                        return this.initiateCommunication(socket)
+                            .then(() => {
+                                return clientConnection.registerResponseCallback(processResponseCallback);
+                            })
+                            .then(() => {
+                                return this.authenticateOnCluster(clientConnection);
+                            }).then(() => {
+                                connectionResolver.resolve(clientConnection);
+                            });
+                    })
+                    .catch((error) => {
+                        connectionResolver.reject(error);
+                    })
+                    .finally(() => {
+                        this.pendingConnections.delete(addressKey);
+                    });
+            });
+
+        return connectionResolver.promise
+            .timeout(this.connectionTimeoutMillis, new HazelcastError(`Connection timed out to address ${address}.`))
+            .finally(() => {
+                this.pendingConnections.delete(addressKey);
+            });
+    }
+
+    public getRandomConnection(): ClientConnection {
+        if (this.isSmartRoutingEnabled) {
+            const member = this.loadBalancer.next();
+            if (member != null) {
+                const connection = this.getConnection(member.uuid);
+                if (connection != null) {
+                    return connection;
+                }
+            }
```

Comment:
Question: don't we need to handle `member == null` case somehow, e.g. by throwing an exception or returning a `null`? Such situation means that either LB class doesn't work as expected, or there are no known members, right?

Reply:
I copied this method from the Java client. Initially, I thought, in such cases we would return a random connection from the active connections. That's way I added the below TODO. Even in this case, I wanted to return a random connection.

Do you think that we should change the current behaviour (also from the Java client)?

Reply:
I've asked this question because of the TODO comment. As `LoadBalancer`'s java/jsdoc allows `next()` method to return a `null`, I think it's fine to handle it in the current way, i.e. by falling back into unisocket behavior and returning a `null` if no active connections are available.